### PR TITLE
[Migrate] Master account analytic default rap

### PR DESCRIPTION
**Pull request by rap-odoo**
<strong>Task:</strong>https://www.odoo.com/web#id=12154&view_type=form&model=project.task&menu_id=3942&action=327
<strong> Pad:</strong>https://pad.odoo.com/p/openerp-project.task-42RUXSFUO3

### CODE REVIEW

File: addons/account_analytic_default/models/account_analytic_default.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from openerp import api, fields, models
+
+
+class AccountAnalyticDefault(models.Model):
+    _name = "account.analytic.default"
+    _description = "Analytic Distribution"
+    _order = "sequence"
```
Comment (atp-odoo) [inline addons/account_analytic_default/models/account_analytic_default.py]: forgot to add _rec_name = "analytic_id" ?

File: addons/account_analytic_default/models/account_analytic_default.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = "account.invoice.line"
+    _description = "Invoice Line"
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        res = super(AccountInvoiceLine, self)._onchange_product_id()
+        rec = self.env['account.analytic.default'].account_get(self.product_id.id, self.invoice_id.partner_id.id, self._uid,
```
Comment (atp-odoo) [inline addons/account_analytic_default/models/account_analytic_default.py]: pass self.env.uid instead of self._uid

File: addons/account_analytic_default/models/account_analytic_default.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                                                               fields.Date.today(), company_id=self.company_id.id)
+        if rec:
+            self.account_analytic_id = rec.analytic_id.id
+        else:
+            self.account_analytic_id = False
```
Comment (atp-odoo) [inline addons/account_analytic_default/models/account_analytic_default.py]: no need to check if and else. not required.  rec is recordset. so if no value then return empty recordset.

File: addons/account_analytic_default/models/product.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    @api.multi
+    def _compute_rules_count(self):
+        self.ensure_one()
+        for product_tmpl_id in self:
+            self.rules_count = sum([p.rules_count for p in product_tmpl_id.product_variant_ids])
```
Comment (atp-odoo) [inline addons/account_analytic_default/models/product.py]: use @api.multi with read_group

File: addons/account_analytic_default/models/sale_order_line.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from openerp import api, fields, models
```
Comment (atp-odoo) [inline addons/account_analytic_default/models/sale_order_line.py]: unused import fields

File: addons/account_analytic_default/models/sale_order_line.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    # Method overridden to set the analytic account by default on criterion match
+    @api.multi
+    def invoice_line_create(self, invoice_id, quantity):
+        self.ensure_one()
+        create_ids = super(SaleOrderLine, self).invoice_line_create(invoice_id, quantity)
+        if not self.ids:
```
Comment (atp-odoo) [inline addons/account_analytic_default/models/sale_order_line.py]: you can directly check with self. if not self : ...

File: addons/account_analytic_default/models/sale_order_line.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+            return create_ids
+        for line in create_ids:
+            rec = self.env['account.analytic.default'].account_get(line.product_id.id, self.order_id.partner_id.id, self.order_id.user_id.id, fields.Date.today())
```
Comment (atp-odoo) [inline addons/account_analytic_default/models/sale_order_line.py]: declare self.env outside loop.

File: addons/account_analytic_default/security/account_analytic_default_security.xml [hunk 8]
```diff
@@ -1,13 +1,11 @@
...
-
-    <record id="analytic_default_comp_rule" model="ir.rule">
-        <field name="name">Analytic Default multi company rule</field>
-        <field name="model_id" ref="model_account_analytic_default"/>
-        <field eval="True" name="global"/>
-        <field name="domain_force">['|',('company_id','=',False),('company_id','child_of',[user.company_id.id])]</field>
-    </record>
-     
+        <record id="analytic_default_comp_rule" model="ir.rule">
+            <field name="name">analytic.default.multi.company.rule</field>
```
Comment (atp-odoo) [inline addons/account_analytic_default/security/account_analytic_default_security.xml]: no need to change name for ir.rule. it's better to keep string.

File: addons/account_analytic_default/views/account_analytic_default_view.xml [hunk 9]
```diff
@@ -74,29 +73,29 @@
             <field name="res_model">account.analytic.default</field>
             <field name="context">{'search_default_product_id': [active_id], 'default_product_id': active_id}</field>
         </record>
-        <record model="ir.ui.view" id="product_form_view_default_analytic_button">
+        <record id="product_form_view_default_analytic_button" model="ir.ui.view">
```
Comment (atp-odoo) [inline addons/account_analytic_default/views/account_analytic_default_view.xml]: split this product and product.template changes in separate view file same like which you did for python model

File: addons/account_analytic_default/views/account_analytic_default_view.xml [hunk 10]
```diff
@@ -74,29 +73,29 @@
...
-                        <field string="Analytic Rules" name="rules_count" widget="statinfo" />
+                        <field name="rules_count" string="Analytic Rules" widget="statinfo" />
                     </button>
                 </div>
             </field>
         </record>
         
-        <record model="ir.ui.view" id="product_template_view_default_analytic_button">
+        <record id="product_template_view_default_analytic_button" model="ir.ui.view">
             <field name="name">product.template.stock.move</field>
```
Comment (atp-odoo) [inline addons/account_analytic_default/views/account_analytic_default_view.xml]: product.template.inherit.form.move

File: addons/account_analytic_default/views/account_analytic_default_view.xml [hunk 11]
```diff
@@ -74,29 +73,29 @@
             <field name="res_model">account.analytic.default</field>
             <field name="context">{'search_default_product_id': [active_id], 'default_product_id': active_id}</field>
         </record>
-        <record model="ir.ui.view" id="product_form_view_default_analytic_button">
+        <record id="product_form_view_default_analytic_button" model="ir.ui.view">
             <field name="name">product.product.stock.move</field>
```
Comment (atp-odoo) [inline addons/account_analytic_default/views/account_analytic_default_view.xml]: product.product.inherit.form.move

File: addons/sale/sale.py [hunk 12]
```diff
@@ -608,7 +608,7 @@ def invoice_line_create(self, invoice_id, qty):
             if not float_is_zero(qty, precision_digits=precision):
                 vals = line._prepare_invoice_line(qty=qty)
                 vals.update({'invoice_id': invoice_id, 'sale_line_ids': [(6, 0, [line.id])]})
-                self.env['account.invoice.line'].create(vals)
+                return self.env['account.invoice.line'].create(vals)
```
Comment (atp-odoo) [inline addons/sale/sale.py]: why this change ?

File: addons/account_analytic_default/models/account_analytic_default.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+class AccountInvoiceLine(models.Model):
+    _inherit = "account.invoice.line"
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        res = super(AccountInvoiceLine, self)._onchange_product_id()
+        rec = self.env['account.analytic.default'].account_get(self.product_id.id, self.invoice_id.partner_id.id, self.env.uid,
+                                                               fields.Date.today(), company_id=self.company_id.id)
+        if rec:
+            self.account_analytic_id = rec.analytic_id.id or ''
```
Comment (mdi-odoo) [inline addons/account_analytic_default/models/account_analytic_default.py]: No need to check the condition, Just self.account_analytic_id = rec.analytic_id is enough .

File: addons/account_analytic_default/models/product.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    @api.multi
+    def _compute_rules_count(self):
+        for product in self:
+            product.rules_count = self.env['account.analytic.default'].search_count([('product_id', '=', product.id)])
```
Comment (mdi-odoo) [inline addons/account_analytic_default/models/product.py]: Get the model instance outside the for loop.

File: addons/account_analytic_default/models/product.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+from odoo import api, fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    @api.multi
+    def _compute_rules_count(self):
```
Comment (mdi-odoo) [inline addons/account_analytic_default/models/product.py]: No need @api.multi and also specify the correct field dependencies of a 'compute' method .

File: addons/account_analytic_default/models/product.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+            product.rules_count = self.env['account.analytic.default'].search_count([('product_id', '=', product.id)])
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    @api.multi
+    def _compute_rules_count(self):
```
Comment (mdi-odoo) [inline addons/account_analytic_default/models/product.py]: No need @api.multi and also specify the correct field dependencies of a 'compute' method .

File: addons/account_analytic_default/models/account_analytic_default.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+class AccountInvoiceLine(models.Model):
+    _inherit = "account.invoice.line"
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        res = super(AccountInvoiceLine, self)._onchange_product_id()
+        rec = self.env['account.analytic.default'].account_get(self.product_id.id, self.invoice_id.partner_id.id, self.env.uid,
+                                                               fields.Date.today(), company_id=self.company_id.id)
+        self.account_analytic_id = rec and rec.analytic_id.id or ''
```
Comment (mdi-odoo) [inline addons/account_analytic_default/models/account_analytic_default.py]: Just self.account_analytic_id = rec.analytic_id is enough .

File: addons/account_analytic_default/models/account_analytic_default.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+            domain += ['|', ('company_id', '=', company_id)]
+        domain += [('company_id', '=', False)]
+        if user_id:
+            domain += ['|', ('user_id', '=', user_id)]
+        domain += [('user_id', '=', False)]
+        if date:
+            domain += ['|', ('date_start', '<=', date), ('date_start', '=', False)]
+            domain += ['|', ('date_stop', '>=', date), ('date_stop', '=', False)]
+        best_index = -1
+        res = False
```
Comment (hmo-odoo) [inline addons/account_analytic_default/models/account_analytic_default.py]: can we initialize  with empty record set ?

File: addons/account_analytic_default/models/product.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    def _compute_rules_count(self):
+        for product_tmpl_id in self:
+            product_tmpl_id.rules_count = sum([p.rules_count for p in product_tmpl_id.product_variant_ids])
```
Comment (hmo-odoo) [inline addons/account_analytic_default/models/product.py]: can we use read_group ?

File: addons/account_analytic_default/models/sale_order_line.py [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    # Method overridden to set the analytic account by default on criterion match
+    @api.multi
+    def invoice_line_create(self, invoice_id, quantity):
+        create_ids = super(SaleOrderLine, self).invoice_line_create(invoice_id, quantity)
+        if not self:
```
Comment (hmo-odoo) [inline addons/account_analytic_default/models/sale_order_line.py]: remove id statement

File: addons/account_analytic_default/models/sale_order_line.py [hunk 21]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
```
Comment (hmo-odoo) [inline addons/account_analytic_default/models/sale_order_line.py]: unused import: fields

File: addons/account_analytic_default/models/sale_order_line.py [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    # Method overridden to set the analytic account by default on criterion match
+    @api.multi
+    def invoice_line_create(self, invoice_id, quantity):
```
Comment (hmo-odoo) [inline addons/account_analytic_default/models/sale_order_line.py]: instead of override invoice_line_create method, override _prepare_invoice_line method

File: addons/account_analytic_default/models/sale_order_line.py (hunk 21 above)
Comment (rap-odoo) [inline addons/account_analytic_default/models/sale_order_line.py]: fields is used in line 16.

File: addons/account_analytic_default/models/product.py [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    def _compute_rules_count(self):
+        product_ids = self.mapped('product_variant_ids').ids
+        product_data = self.env['account.analytic.default'].read_group([('product_id', 'in', product_ids)], ['product_id'], ['product_id'])
+        for data in product_data:
+            self.rules_count += int(data['product_id_count'])
+
+    @api.multi
+    def action_view_rules(self):
```
Comment (hmo-odoo) [inline addons/account_analytic_default/models/product.py]: put ensure_one()

**Comment by mart-e** [comment]
@rap-odoo you are right the stock code was no longer useful but please, explain your commits ! "Remove the unnecessary code." is very suspicious commit message. **Why**, not **what**.

**Comment by mart-e** [comment]
@rap-odoo spliting in different files is ok for huge files but not needed here, just making uncessary commits and breaking blame

File: addons/account_analytic_default/models/product.py [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    def _compute_rules_count(self):
+        product_ids = self.mapped('product_variant_ids').ids
+        product_data = self.env['account.analytic.default'].read_group([('product_id', 'in', product_ids)], ['product_id'], ['product_id'])
```
Comment (mart-e) [inline addons/account_analytic_default/models/product.py]: this looks like a change of behaviour or an optimisation, in second case should be in separated commit, not in migration code

File: addons/account_analytic_default/models/product.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+            product.rules_count = Analytic.search_count([('product_id', '=', product.id)])
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    rules_count = fields.Integer(compute='_compute_rules_count', string='# Analytic Rules')
+
+    def _compute_rules_count(self):
+        product_ids = self.mapped('product_variant_ids').ids
```
Comment (mart-e) [inline addons/account_analytic_default/models/product.py]: I don't think that makes any effect other than reducing the readability.
```
>>> timeit.timeit("lambda: x.mapped('product_variant_ids').ids", number=10000000)
0.7764549255371094
>>> timeit.timeit("lambda: [p.product_variant_ids.ids for p in x]", number=10000000)
0.7627608776092529
```

File: addons/account_analytic_default/models/product.py [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+        product_data = self.env['account.analytic.default'].read_group([('product_id', 'in', product_ids)], ['product_id'], ['product_id'])
+        for data in product_data:
+            self.rules_count += int(data['product_id_count'])
```
Comment (mart-e) [inline addons/account_analytic_default/models/product.py]: `+=` ? Cool, the rules_count increase each time I evaluate this method.
`self`? so the same count is for every record?

Comment (rim-odoo) [inline addons/account_analytic_default/models/product.py]: <iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/sC75aU47GRk?rel=0" frameborder="0" allowfullscreen></iframe>
